src/highdicom/seg/sop.py

```diff
@@ -1932,29 +1905,287 @@ def _omit_empty_frames(
+        if pixel_array.dtype in (np.float_, np.float32, np.float64):
+            # Based on the previous checks and casting, if we get here the
+            # output is a FRACTIONAL segmentation Floating-point numbers must
+            # be mapped to 8-bit integers in the range [0,
+            # max_fractional_value].
+            if pixel_array.ndim == 4:
+                segment_array = pixel_array[:, :, :, segment_number - 1]
+            else:
+                segment_array = pixel_array
+            segment_array = np.around(
+                segment_array * float(max_fractional_value)
+            )
+            segment_array = segment_array.astype(np.uint8)
+        else:
+            if pixel_array.ndim == 3:
+                # "Label maps" that must be converted to binary masks.
+                if number_of_segments == 1:
+                    # We wish to avoid unnecessary comparison or casting
+                    # operations here, for efficiency reasons. If there is only
+                    # a single segment, the label map pixel array is already
+                    # correct
+                    if pixel_array.dtype != np.uint8:
+                        segment_array = pixel_array.astype(np.uint8)
+                    else:
+                        segment_array = pixel_array
+                else:
+                    segment_array = (
+                        pixel_array == segment_number
+                    ).astype(np.uint8)
+            else:
+                segment_array = pixel_array[:, :, :, segment_number - 1]
+                if segment_array.dtype != np.uint8:
+                    segment_array = segment_array.astype(np.uint8)
+
+            # It may happen that a binary valued array is passed that should be
+            # stored as a fractional segmentation. In this case, we also need
+            # to stretch pixel values to 8-bit unsigned integer range by
+            # multiplying with the maximum fractional value.
+            if segmentation_type == SegmentationTypeValues.FRACTIONAL:
+                # Avoid an unnecessary multiplication operation if max
+                # fractional value is 1
+                if int(max_fractional_value) != 1:
+                    segment_array *= int(max_fractional_value)
+
+        return segment_array
+
+    @staticmethod
+    def _get_dimension_index_values(
+        plane_index: int,
+        dimension_position_values: List[np.ndarray],
+        plane_position_values: np.ndarray,
+        has_ref_frame_uid: bool,
+        coordinate_system: Optional[CoordinateSystemNames],
+    ) -> List[int]:
+        """Get dimension index values for a frame.
+
+        Parameters
+        ----------
+        plane_index: int
+            Index of the plane in the sorted planes list.
+        dimension_position_values: List[numpy.ndarray]
+            Relative locations of each plane.
+        plane_position_values: numpy.ndarray
+            Plane positions of each plane.
```

Comment:
Agreed. Trying to actually understand what all this does (rather than naively refactoring it keeping the logic the same) led me down quite a rabbit hole. I ended up decided that the current implementation, though correct, is needlessly complex as it maintains two lists of indices and different arrays sorted according to them. I refactored things to keep everything indexing the original list. I also changed some variable and (private) parameter names that were confusing in my opinion, and updated these docstrings. See 287047e